File: app/lib/domain/models/intelligence.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'intelligence.freezed.dart';
part 'intelligence.g.dart';

/// The commute this user is most likely making right now, learned from
/// their own journey history — not a generic "next scheduled train" lookup.
/// `confidence` (0..1) and `sampleSize` are always present so the UI can be
/// honest about how sure a prediction really is.
@freezed
class CommutePrediction with _$CommutePrediction {
  const factory CommutePrediction({
    required String originStopId,
    required String originName,
    required String destinationStopId,
    required String destinationName,
    String? routeId,
    String? routeLongName,
    required DateTime predictedDepartureAt,
    double? predictedDurationSeconds,
    int? recommendedCoach,
    String? recommendedExitName,
    required double confidence,
    required int sampleSize,
    required String basis,
  }) = _CommutePrediction;

  factory CommutePrediction.fromJson(Map<String, dynamic> json) =>
      _$CommutePredictionFromJson(json);
}

/// Typical delay for a route around an hour of day, from historical
/// completed-journey durations vs. GTFS-scheduled duration.
@freezed
class DelayEstimate with _$DelayEstimate {
  const factory DelayEstimate({
    required String routeId,
    int? directionId,
    required int hourOfDay,
    required double expectedDelaySeconds,
    required double confidence,
    required int sampleSize,
  }) = _DelayEstimate;

  factory DelayEstimate.fromJson(Map<String, dynamic> json) =>
      _$DelayEstimateFromJson(json);
}

/// One scored route option inside a smart-recommendation bundle.
@freezed
class RouteRecommendation with _$RouteRecommendation {
  const factory RouteRecommendation({
    required String preference,
    required double travelSeconds,
    required int interchangeCount,
    required double walkingDistanceM,
    required double delayAdjustedSeconds,
    required List<String> reasons,
  }) = _RouteRecommendation;

  factory RouteRecommendation.fromJson(Map<String, dynamic> json) =>
      _$RouteRecommendationFromJson(json);
}

/// The synthesised "what should I do" bundle for an origin/destination pair.
@freezed
class SmartRecommendation with _$SmartRecommendation {
  const factory SmartRecommendation({
    required String originStopId,
    required String destinationStopId,
    DateTime? bestDepartureAt,
    RouteRecommendation? bestRoute,
    required List<RouteRecommendation> alternatives,
    int? recommendedCoach,
    String? recommendedExitName,
    required bool leastCrowdedAvailable,
  }) = _SmartRecommendation;

  factory SmartRecommendation.fromJson(Map<String, dynamic> json) =>
      _$SmartRecommendationFromJson(json);
}

/// A place role inferred from journey history — a suggestion to offer the
/// user (e.g. pre-filling a Favourites label), never written on their
/// behalf. `role` is `"home"` or `"weekday_anchor"`: movement data alone
/// can't tell "Office" from "College", so the app just calls it a regular
/// weekday destination and lets the user name it.
@freezed
class InferredPlace with _$InferredPlace {
  const factory InferredPlace({
    required String stopId,
    required String stopName,
    required String role,
    required double confidence,
    required int sampleSize,
    required String rationale,
  }) = _InferredPlace;

  factory InferredPlace.fromJson(Map<String, dynamic> json) =>
      _$InferredPlaceFromJson(json);
}

File: src/metropulse/domain/intelligence.py
"""Domain values for Metro Intelligence.

Every prediction here is derived from GTFS schedules plus the user's own
journey history — never a black-box model. Each carries an explicit
``confidence`` (0..1, scaling with sample size) and, where relevant, the
``sample_size`` it's based on, so callers can be honest about how sure a
prediction really is rather than presenting a guess as a fact.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from enum import Enum


@dataclass(frozen=True, slots=True)
class CommutePrediction:
    """The commute this user is most likely making right now, learned from
    their own journey history — not a generic "next scheduled train" lookup.
    """

    origin_stop_id: str
    origin_name: str
    destination_stop_id: str
    destination_name: str
    route_id: str | None
    route_long_name: str | None
    predicted_departure_at: datetime
    predicted_duration_seconds: float | None
    recommended_coach: int | None
    recommended_exit_name: str | None
    confidence: float
    sample_size: int
    basis: str


@dataclass(frozen=True, slots=True)
class DelayEstimate:
    """Typical delay for a route around an hour of day, from completed
    journeys' actual duration vs. the GTFS-scheduled duration for the same
    trip. Positive means slower than scheduled, negative means faster.
    """

    route_id: str
    direction_id: int | None
    hour_of_day: int
    expected_delay_seconds: float
    confidence: float
    sample_size: int


@dataclass(frozen=True, slots=True)
class RouteRecommendation:
    """One scored route option inside a smart-recommendation bundle."""

    preference: str
    travel_seconds: float
    interchange_count: int
    walking_distance_m: float
    delay_adjusted_seconds: float
    reasons: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class SmartRecommendation:
    """The synthesised "what should I do" bundle for an origin/destination.

    ``least_crowded_available`` is always False today: per-coach occupancy
    feeds the coach recommendation already, but route-level crowd-density
    scoring (choosing between whole routes by expected crowding) isn't
    modelled yet — this field is the explicit, honest hook for that future
    capability rather than a fabricated score.
    """

    origin_stop_id: str
    destination_stop_id: str
    best_departure_at: datetime | None
    best_route: RouteRecommendation | None
    alternatives: tuple[RouteRecommendation, ...]
    recommended_coach: int | None
    recommended_exit_name: str | None
    least_crowded_available: bool


class PlaceRole(str, Enum):
    """A role inferred for a place from movement patterns, not user input.

    Movement data alone can't tell "Office" from "College" — both are simply
    the place a user travels to and from on a regular weekday schedule. So
    this models exactly two roles: ``HOME`` (where trips most often start)
    and ``WEEKDAY_ANCHOR`` (where they most often go on weekdays from home).
    The UI is expected to let the user confirm or relabel the real name.
    """

    HOME = "home"
    WEEKDAY_ANCHOR = "weekday_anchor"


@dataclass(frozen=True, slots=True)
class InferredPlace:
    """A place role inferred from a user's own journey history.

    This is a suggestion, not a fact written on the user's behalf: callers
    should offer it for confirmation (e.g. pre-filling a Favourites label)
    rather than silently overwriting anything the user already set.
    """

    stop_id: str
    stop_name: str
    role: PlaceRole
    confidence: float
    sample_size: int
    rationale: str
